Encoding malformed characters

4513 opened the pull request:
*No collected data available*

While profiling an application, malformed data were sent to a saver (both, Upload and File) that called `json_encode` with the provided data.  
That resulted into an encoding error, returning false that was casted into an empty string.  

Because of that, no collected data was available at the end of the request.

The proposed change uses [JSON_INVALID_UTF8_IGNORE](https://www.php.net/manual/en/json.constants.php#constant.json-invalid-utf8-ignore) constant which removes all invalid characters, making the encoding possible.

The constant is available since PHP 7.2, that is why a 'check' is made before the encoding. If the PHP has the constant, it will use it.

---
Even though the change removes a word/key from the resulted output, the change does not have any undesirable changes because the previous (current) behavior does not generate the output at all.

glensc commented:
Can you refactor this using a single constant `PHP_VERSION_ID` check and setting $options rather than multiple json_encode calls? (default is `$flags = 0` - https://www.php.net/json_encode)

Example:
- https://github.com/zf1s/zf1/blob/ee72a882b1d1ec39c6db761314d8055c1cf67501/packages/zend-validate/library/Zend/Validate/StringLength.php#L202-L204

glensc commented:
also, you should never create PR from default branch (always create feature branch), making changes for the PR would mess up your default branch and vice versa. you might want to keep them separate.

glensc commented:
also, while at it, fix Mlaformed -> malformed typo in commit message.

glensc commented on src/Saver/FileSaver.php:
```diff
@@ -19,8 +19,20 @@ public function isSupported()
+            : json_encode($data);
 
-        return file_put_contents($this->file, $json . PHP_EOL, FILE_APPEND);
+        return $result === false ? '' : $result;
```
why put empty line to file? maybe skip whole write then?

glensc commented on src/Saver/FileSaver.php:
```diff
@@ -19,8 +19,20 @@ public function isSupported()
+     */
+    private function encodeData(array $data)
+    {
+        $result = PHP_MAJOR_VERSION >= 7 && PHP_MINOR_VERSION >= 2
```
this check fails with php 8.0, use PHP_VERSION_ID instead

4513 commented:
* changed to use `PHP_VERSION_ID` instead of the previous two constants
* changed to rather setting `$option` than calling `json_econde` differently
* added condition to check that the result of json_encode is valid. If false is returned, the `save` is immediately stopped with return 'false' (no saving into file or sending a request to upload the json)

Please, note that I created a new branch as noted on which the proposed changes are made, including fixed typo in commit message.

If you are okay, with that, I will create a new request from that branch. Meanwhile, marking this PR as draft, avoiding the merging

glensc commented:
if you're able to force push those changes to your `main` branch (this branch source branch) then you don't need to make new pr. if you're unable figure out how to make force push, just add commits to this branch and I'll squash merge (then your commit messages will be lost, will be replaced what I fill).

i prefer if we can continue in this pr, so discussions don't get spread out between multiple pr's.

4513 commented:
`main` branch has been force pushed and contains all the changes now

4513 commented:
thank you 🙂
